hub-saml/src/main/java/uk/gov/ida/saml/core/security/AssertionsDecryptersForValidatedResponse.java

```diff
@@ -0,0 +1,52 @@
+package uk.gov.ida.saml.core.security;
+
+import org.opensaml.saml.saml2.core.Assertion;
+import org.slf4j.LoggerFactory;
+import org.slf4j.event.Level;
+import uk.gov.ida.saml.hub.validators.response.common.UnDecryptedValidatedResponseCounter;
+import uk.gov.ida.saml.security.AssertionDecrypter;
+import uk.gov.ida.saml.security.exception.SamlFailedToDecryptException;
+
+import java.util.List;
+
+public class AssertionsDecryptersForValidatedResponse { 
+    private final List<AssertionDecrypter> assertionDecrypters;
+    
+    public AssertionsDecryptersForValidatedResponse(List<AssertionDecrypter> decrypters){
+        this.assertionDecrypters = decrypters;
+    }
+    
+    public List<Assertion> decryptedAssertions(String className, UnDecryptedValidatedResponseCounter unDecryptedValidatedResponseCounter){
+        var decryptedAssertions =  getDecryptedAssertions(className, unDecryptedValidatedResponseCounter);
+        
+        if (decryptedAssertions == null) { 
+            throw new SamlFailedToDecryptException(
+                    String.format(
+                            "%s could not decrypt assertions from %s with any of the decrypters",
+                            className,
+                            unDecryptedValidatedResponseCounter.getValidatedResponse().getIssuer().getValue()
+                    ), Level.ERROR
+            );
+        }
+        
+        return decryptedAssertions;
+    }
+    
+    private List<Assertion> getDecryptedAssertions(String className, UnDecryptedValidatedResponseCounter unDecryptedValidatedResponseCounter) {
+        for (AssertionDecrypter assertionDecrypter : assertionDecrypters) {
+            try {
+                return assertionDecrypter.decryptAssertions(unDecryptedValidatedResponseCounter.getValidatedResponse());
+            } catch(SamlFailedToDecryptException e) {
+                String issuer = unDecryptedValidatedResponseCounter.getValidatedResponse().getIssuer().getValue();
+                unDecryptedValidatedResponseCounter.increment(issuer);
+                LoggerFactory.getLogger(className)
+                             .warn(
+                                     String.format("%s failed to decrypt assertions from %s with one of the decrypters",
+                                     className,
+                                     issuer), e
+                             );
+            }
+        }
+        return null;
```

Comment:
Could we consider returning an optional here?

Reply:
I'm going to weigh in here and suggest an empty list as the best practice where lists that might be empty are concerned. `return Collections.emptyList();` and `if(decryptedAssertions.isEmpty())` can be dropped in where the nulls are used.

Reply:
@JohnWattsGDS I actually tried that and it broken some other tests, I have refactored this entirely, more inline with logic suggested by @CharlesIC